text/0000-write-pointers.md

```diff
@@ -0,0 +1,468 @@
+
+```Rust
+fn maybe_init(maybe: Option<&uninit T>) { ... }
+fn init(ref_to_write: &mut &uninit T) { ... }
+struct Temp { a: &uninit Foo }
+```
+
+#### Conditional Initialization
+
+One restriction to `&uninit T` is that we cannot conditionally initialize a value. For example, none of these are allowed.
+```Rust
+let x: Foo;
+let condition = ...;
+
+if condition {
+    x.a = 12; // Error: Conditional partial initialization is not allowed
+}
+```
+```Rust
+let mut x: Foo;
+let condition = ...;
+
+while condition {
+    x.a = 12; // Error: Conditional partial initialization is not allowed
+}
+```
+```Rust
+let mut x: Foo;
+
+for ... {
+    x.a = 12; // Error: Conditional partial initialization is not allowed
+}
+```
+Because if we do, then we can't gaurentee that the value is in fact initialized.
+
+Note, that this is not conditionally initializing `x.e`, because by the end of the `if-else` block, `x.e` is guaranteed to be initialized.
+```Rust
+let x: Bar;
+
+x.d = 10;
+
+if { ... any condition ... } {
+    x.e = 1.0;
+} else {
+    x.e = 0.0;
+}
+```
+
+### Using partially initialized variables
+
+```Rust
+let x: Bar;
+x.d = 2;
+
+// This is fine, we know that x.d is initialized
+x.d.pow(4);
+if x.d == 16 {
+    x.e = 10.0;
+} else {
+    x.e = 0.0;
+}
+// This is fine, we know that x is initialized
+assert_eq!(x.e, 10.0);
+```
+
+### Functions and closures
+
+You can accept `&uninit T` as arguments to a function or closure.
+
+```Rust
+fn init_foo(foo: &uninit Foo) { ... }
+let init_bar = |bar: &uninit Bar| { ... }
+```
+
+But if you do accept a `&uninit T` argument, you must write to it before returning from the function or closure.
+
+```Rust
+fn valid_init_bar_v1(bar: &uninit Bar) {
+    bar.d = 10;
+    bar.e = 2.7182818;
+}
+fn valid_init_bar_v2(bar: &uninit Bar) {
+    // you must dereference if you write directly to a &uninit T
+    // This still does not drop the old value of bar
+    *bar = Bar { d: 10, e: 2.7182818 };
+}
+fn invalid_init_bar_v1(bar: &uninit Bar) {
+    bar.d = 10;
+    // Error, bar is not completely initialized (Bar.e is not initialized)
+}
+
+fn invalid_init_bar_v2(bar: &uninit Bar) {
+    bar.d = 10;
+    if bar.d == 9 {
+        return; // Error, bar is not completely initialized (Bar.e is not initialized)
+    }
+    bar.e = 10.0;
+}
+```
+
+If a closure captures a `&uninit T`, then it becomes a `FnOnce`, because of the write semantics, the destructors will not be run the first time.
+
+```Rust
+let x: Foo;
+
+let init = || x.a = 12; // init: FnOnce()  -> ()
+```
+
+**Note on Panicky Functions:**
+If a function panics, then all fields initialized in that function will be dropped. No cross-function analysis will be done.
+
+## `&out T`
+
+Using `&out T`, we can directly initialize a value and guarantee to write only behavior. \
+That would add give a memory location to write to directly instead of relying on move-elimination optimizations.
```

Comment:
The two code blocks under this paragraph only use `&uninit`, never `&out`.  Should there be something else there instead?